### PR TITLE
doc: expand the `<:` doc string

### PR DESCRIPTION
**Pull request by nsajko**
Clear up some things, state some expected properties and limitations, add cross-references, add more examples.

### CODE REVIEW

File: base/operators.jl [hunk 1]
```diff
@@ -3,10 +3,27 @@
 ## types ##
 
 """
-    <:(T1, T2)
+```julia
```
Comment (Keno) [inline base/operators.jl:6]: The signature is not ordinarily a code block

File: base/operators.jl [hunk 2]
```diff
@@ -3,10 +3,27 @@
...
+<:(L, R)
+```
+
+The subtyping predicate: test whether the type `L` may be a subtype of `R`.
+
+For any type `A` and any type `B`, `A` subtypes `B` if and only if any value
```
Comment (Keno) [inline base/operators.jl:12]: This condition is sufficient but not necessary. As written this sounds like this is the definition of subtyping, but sometimes julia gets it wrong. That's not really the case.

File: base/operators.jl [hunk 3]
```diff
@@ -3,10 +3,27 @@
...
+Julia is sometimes not able to prove a subtyping relation between two types, even
+if such a relation holds. In such a case the result of `S <: T` will be `false`
+even if `S` subtypes `T`. Even so, `<:` is expected to be reflexive and transitive.
+
+See also: [the Manual page on types](@ref man-types), [`Union{}`](@ref),
+[`Any`](@ref), [`isa`](@ref).
 
-Subtype operator: returns `true` if and only if all values of type `T1` are
-also of type `T2`.
+Julia's subtyping relation is the topic of the paper *Julia Subtyping: A Rational
```
Comment (Keno) [inline base/operators.jl:23]: Ordinary style is for citations to go at the end of the docstring.

File: base/operators.jl (hunk 2 above)
Comment (nsajko) [reply base/operators.jl:12]: That's kind of the point of this PR. The previous version of the doc string didn't mention that Julia can get it wrong, so in this PR I try separating the definition of subtyping from the definition of `<:`.

File: base/operators.jl (hunk 1 above)
Comment (nsajko) [reply base/operators.jl:6]: I just changed one code block syntax to another code block syntax. The Markdown [docs](https://docs.julialang.org/en/v1/stdlib/Markdown/) mention that the syntax I used should be preferred to the existing syntax:

> "Fenced" code blocks, as shown in the last example, should be preferred over indented code blocks since there is no way to specify what language an indented code block is written in.

Should I disregard that note?

File: base/operators.jl (hunk 2 above)
Comment (Keno) [reply base/operators.jl:12]: Yes, but that's the wrong phrasing. The definition of subtyping is the complicated set of rules that define it. This definition implies (in one direction, so no `if and only if`), the value constraint, but types are not defined by the set of values that are of that type. Taking the "sets of values" view of subtyping can be a good mental model, but it's an approximation.

File: base/operators.jl (hunk 1 above)
Comment (Keno) [reply base/operators.jl:6]: The first line of the docstring is a bit special as it's the signature, which is sort of an underspecified pseudo-julia syntax. It's not really supposed to be a code block, it just happens to be rendered that way.

File: base/operators.jl (hunk 3 above)
Comment (tpapp) [reply base/operators.jl:23]: As much as I like that paper, I am not sure it is a good idea to cite it in the Julia docs. It addressed a previous state of the type system and many suggestions in the paper have been acted upon.

Comment (nsajko) [reply base/operators.jl:23]: True, it's not up-to-date anymore for current Julia. I think you may be right, but there's some things in that paper that should maybe appear in the Julia docs, like Table 1, *Julia subtyping compared*.

File: base/operators.jl [hunk 4]
```diff
@@ -3,10 +3,27 @@
...
-    <:(T1, T2)
+```julia
+<:(L, R)
+```
+
+The subtyping predicate: test whether the type `L` may be a subtype of `R`.
+
+For any type `A` and any type `B`, `A` subtypes `B` if and only if any value
+belonging to `A` is also of type `B`. Subtyping is reflexive and transitive,
+making it a *preorder*.
```
Comment (nsajko) [inline base/operators.jl:14]: `<:` is also antisymmetric, by definition of `==`, which is defined in terms of `<:`. So (if we actually had transitivity, see #53019 and #53021) we could actually use the stronger label of *partial order*

File: base/operators.jl [hunk 5]
```diff
@@ -16,9 +25,27 @@ true
 julia> Vector{Int} <: AbstractArray
 true
 
-julia> Matrix{Float64} <: Matrix{AbstractFloat}
+julia> Matrix{Float64} <: Matrix{AbstractFloat}  # invariance
 false
+
+julia> Tuple{Float64} <: Tuple{AbstractFloat}    # contravariance
```
Comment (jakobnissen) [inline base/operators.jl:31]: This is covariance, not contravariance IIUC

File: base/operators.jl [hunk 6]
```diff
@@ -5,9 +5,18 @@
...
     <:(T1, T2)
 
-Subtype operator: returns `true` if and only if all values of type `T1` are
+Subtype predicate: returns `true` if and only if all values of type `T1` are
 also of type `T2`.
 
+Julia is sometimes not able to prove a subtyping relation that would ideally
+hold between two types. When that happens, `<:` returns `false`.
+
+The `<:` predicate is expected (modulo bugs) to behave as a partial order. I.e.,
```
Comment (jakobnissen) [inline base/operators.jl:14]: Does it really make sense to hedge against bugs here? Surely, any and all documented behaviour ever are always "if there isn't any bugs"

File: base/operators.jl [hunk 7]
```diff
@@ -5,9 +5,18 @@
 """
     <:(T1, T2)
 
-Subtype operator: returns `true` if and only if all values of type `T1` are
+Subtype predicate: returns `true` if and only if all values of type `T1` are
 also of type `T2`.
 
+Julia is sometimes not able to prove a subtyping relation that would ideally
```
Comment (adienes) [inline base/operators.jl:11]: the word "ideally" here feels a little under-specified. which ideal? I would reword this to be a little more direct with less implications about some hypothetical mental model of types: something more like

> Subtype predicate: returns `true` if and only if all values of type `T1` are
also of type `T2`.

> This operator returns `true` when inference is able to prove a subtyping relationship, and may return `false` otherwise.


aka focusing on what it _does_ do not what it doesn't

**Review by fingolfin — APPROVED** [review]
Looks good to me, but would be good to know if @Keno is happy now

**Comment by nsajko** [comment]
bump

**Review by vtjnash — CHANGES_REQUESTED** [review]

File: base/operators.jl [hunk 8]
```diff
@@ -5,8 +5,26 @@
 """
     <:(T1, T2)
 
-Subtype operator: returns `true` if and only if all values of type `T1` are
-also of type `T2`.
+Subtyping relation, defined between two types. The returned value is of type
+`Bool`.
```
Comment (vtjnash) [inline base/operators.jl:9]: style nit: we do not write this in text, but may put it as a return type annotation on the signature above if relevant
```suggestion
Subtyping relation, defined between two types
```

File: base/operators.jl [hunk 9]
```diff
@@ -5,8 +5,26 @@
...
+
+A type `S` is said to be a *subtype* of a type `T` if and only if we have `S <: T`.
```
Comment (vtjnash) [inline base/operators.jl:11]: This is a tautology. I am not sure it helps to define `<:` in terms of itself.

Comment (nsajko) [reply base/operators.jl]: The idea was more to define what "subtype" means in Julia in terms of `<:`. Not sure if it's better now.

File: base/operators.jl [hunk 10]
```diff
@@ -16,9 +32,25 @@ true
...
+true
+
+julia> Union{} <: Int  # The bottom type, `Union{}`, subtypes each type.
+true
+
+julia> Union{} <: Float32 <: AbstractFloat <: Real <: Number <: Any  # Chaining
+true
 ```
+
+The `<:` syntax is also used in other senses:
```
Comment (fingolfin) [inline base/operators.jl:48]:
```suggestion
The `<:` operator is also used in other senses:
```

File: base/operators.jl [hunk 11]
```diff
@@ -16,9 +32,25 @@ true
...
+
+* to specify the lower and upper bound on a parameter of a [`UnionAll`](@ref) type
```
Comment (fingolfin) [inline base/operators.jl:50]: Maybe also mention its use in `where` clauses?

**Comment by fingolfin** [comment]
@nsajko thanks for the improvements, I like it! I left two more minor suggestions.

**Comment by fingolfin** [comment]
I believe this is ready for merging but let's see if @vtjnash is happy with it now :-)

**Comment by vtjnash** [comment]
I made a couple more style and linguistic updates. I think this is accurate and useful now

**Review by vtjnash — APPROVED** [review]

**Comment by nsajko** [comment]
merge?
